### PR TITLE
Improve implementation of classloader caching

### PR DESCRIPTION
**Pull request by retronym**
- Fix logic error in reference counted closing
  - Add a new policy to cache regardless of timestamps
  - When encountering non file-based classpath entries, be optimistic
    rather than disabling caching.

### CODE REVIEW

File: src/compiler/scala/tools/nsc/classpath/ZipAndJarFileLookupFactory.scala [hunk 1]
```diff
@@ -195,25 +197,55 @@ final class FileBasedCache[T] {
...
       }
     }
   }
-  private val cache = collection.mutable.Map.empty[Seq[Path], Entry]
 
-  def getOrCreate(paths: Seq[Path], create: () => T, closeableRegistry: CloseableRegistry): T = cache.synchronized {
+  def checkCacheability(urls: Seq[URL], checkStamps: Boolean, disableCache: Boolean): Either[String, Seq[java.nio.file.Path]] = {
+    import scala.reflect.io.{AbstractFile, Path}
+    val urlsAndFiles = urls.filterNot(_.getProtocol == "jrt").map(u => u -> AbstractFile.getURL(u))
+    val nonJarZips = urlsAndFiles.filter(_._2 == null).filter(p => Jar.isJarOrZip(p._2.file))
```
Comment (mkeskells) [inline src/compiler/scala/tools/nsc/classpath/ZipAndJarFileLookupFactory.scala]: only used conditionally - can be be moved to the block starting at line 242

**Review by mkeskells — CHANGES_REQUESTED** [review]

File: src/compiler/scala/tools/nsc/classpath/ZipAndJarFileLookupFactory.scala [hunk 2]
```diff
@@ -195,43 +197,99 @@ final class FileBasedCache[T] {
...
-  def getOrCreate(paths: Seq[Path], create: () => T, closeableRegistry: CloseableRegistry): T = cache.synchronized {
-    val stamps = paths.map { path =>
-      val attrs = Files.readAttributes(path, classOf[BasicFileAttributes])
-      val lastModified = attrs.lastModifiedTime()
-      // only null on some platforms, but that's okay, we just use the last modified timestamp as our stamp
-      val fileKey = attrs.fileKey()
-      Stamp(lastModified, fileKey)
+  def checkCacheability(urls: Seq[URL], checkStamps: Boolean, disableCache: Boolean): Either[String, Seq[java.nio.file.Path]] = {
+    import scala.reflect.io.{AbstractFile, Path}
+    lazy val urlsAndFiles = urls.filterNot(_.getProtocol == "jrt").map(u => u -> AbstractFile.getURL(u))
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/classpath/ZipAndJarFileLookupFactory.scala:238]: Do we need null checks here, similar to the `hasNullURL` snippet that's been removed?

Comment (adriaanm) [reply src/compiler/scala/tools/nsc/classpath/ZipAndJarFileLookupFactory.scala:238]: ok it looks like that's covered below by checking `file == null`

File: src/compiler/scala/tools/nsc/classpath/ZipAndJarFileLookupFactory.scala [hunk 3]
```diff
@@ -195,43 +197,99 @@ final class FileBasedCache[T] {
   private case class Stamp(lastModified: FileTime, fileKey: Object)
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/classpath/ZipAndJarFileLookupFactory.scala:198]: let's document `fileKey` could be `null`

Comment (retronym) [reply src/compiler/scala/tools/nsc/classpath/ZipAndJarFileLookupFactory.scala:198]: I've updated the code to use a sentinel rather than a null.

**Review by adriaanm — APPROVED** [review]

**Pull request merged**